### PR TITLE
[outdated v0] tts: add e.g. with fastspeech2 and wavegrad

### PR DESCRIPTION
**Pull request by DongyaoZhu**

### CODE REVIEW

File: examples/fastspeech2/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# FastSpeech2
+
+MindSpore implementation of [FastSpeech2](https://arxiv.org/abs/2006.04558), a text-to-melspectrogram model for text-to-speech systems.
```
Comment (vigo999) [inline examples/fastspeech2/README.md:3]: remove MindSpore implementation of

File: examples/fastspeech2/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+## **Dependencies**
+
+1. `pip install -r requirements.txt`
+2. Install [MindSpore](https://www.mindspore.cn/install).
+3. (Optional) Install `mpirun` for distributed training.
+
+## Generate from your data
+
+`python recipes/LJSpeech/tts/fastspeech2/generate.py --restore fastspeech2_160k_en_mel128.ckpt --text "hello this is a test sentence"`
```
Comment (vigo999) [inline examples/fastspeech2/README.md:15]: please correct scripts path

File: examples/fastspeech2/README.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+Put the `TextGrid` to be under your path given in `data_path` in `fastspeech2.yaml`.
+
+#### 0.2 Preprocess
+
+Preprocess data to get a "_wav.npy" and "_feature.npy" for each ".wav" file in your dataset folder. Set your `data_path` and
+`manifest_path` in `fastspeech2.yaml`. You can now run the following command:
+
+```python
+python recipes/LJSpeech/tts/fastspeech2/preprocess.py
```
Comment (vigo999) [inline examples/fastspeech2/README.md:45]: please correct scripts path

File: examples/fastspeech2/README.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+### Step 1 (Train)
+
+#### 1.1 Train on local server
+
+Set up device information:
+```
+export MY_DEVICE=Ascend # options: [Ascend, GPU]
+```
+
+Other training and model parameters can be set in `recipes/LJSpeech/tts/fastspeech2/fastspeech2.yaml`.
```
Comment (vigo999) [inline examples/fastspeech2/README.md:57]: please correct scripts path

File: examples/fastspeech2/README.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+python recipes/LJSpeech/tts/fastspeech2/preprocess.py
+```
+
+### Step 1 (Train)
+
+#### 1.1 Train on local server
+
+Set up device information:
+```
+export MY_DEVICE=Ascend # options: [Ascend, GPU]
```
Comment (vigo999) [inline examples/fastspeech2/README.md:54]: remove this setting. 
user can set target device via train.py args

File: examples/fastspeech2/README.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+Other training and model parameters can be set in `recipes/LJSpeech/tts/fastspeech2/fastspeech2.yaml`.
+
+Train on 1 card:
+```shell
+export DEVICE_ID=0
+nohup python train.py --device_target $MY_DEVICE > train_single.log &
```
Comment (vigo999) [inline examples/fastspeech2/README.md:62]: remove nohup and log info.
python train.py --device_target Ascend

File: examples/fastspeech2/README.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+stay tuned
+```
+
+### Implementation details
+
+
+### Acknowlegements
+
+Some repositories that inspired this implementation:
+- [MindAudio](https://github.com/mindspore-lab/mindaudio)
```
Comment (vigo999) [inline examples/fastspeech2/README.md:76]: remove mindaudio ref

File: examples/fastspeech2/fastspeech2.yaml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+energy:
+    feature: 'frame_level'
+    normalization: False
+    energy_min: 0.0436493270
+    energy_max: 494.149902
+
+data_path: 'data/LJSpeech-1.1'
+# manifest_path: this file is generated, not the metadata.csv
+manifest_path: 'data/ljspeech_manifest.csv'
+lexicon_path: 'recipes/text/librispeech-lexicon.txt'
```
Comment (vigo999) [inline examples/fastspeech2/fastspeech2.yaml:36]: please upload lexicon txt to examples/fastspeech2 folder and correct this path

File: examples/fastspeech2/generate.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    print("batchs:", len(batchs))
+    for i, batch in enumerate(batchs):
+        output = model(p_control=1.0, e_control=1.0, d_control=1.0, **batch)
+        mel_prediction = output["mel_predictions"].asnumpy()
+        np.save("msfs2_%s.npy" % i, mel_prediction)
+
+
+def main():
+    args = parse_args()
+    hps = mindaudio.load_config(args.config)
```
Comment (vigo999) [inline examples/fastspeech2/generate.py:107]: mindaudio load config function has been removed. please upload this scripts to examples/config.py

from pprint import pformat
import numpy as np
import yaml

class Config:
    def __init__(self, cfg_dict):
        for k, v in cfg_dict.items():
            if isinstance(v, (list, tuple)):
                setattr(self, k, [Config(x) if isinstance(x, dict) else x for x in v])
            else:
                setattr(self, k, Config(v) if isinstance(v, dict) else v)

    def __str__(self):
        return pformat(self.__dict__)

    def __repr__(self):
        return self.__str__()


def load_config(path):
    with open(path) as stream:
        hps = yaml.load(stream, yaml.Loader)
    hps = Config(hps)
    return hps

File: examples/fastspeech2/generate.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    ms.context.set_context(mode=mode, device_target=args.device_target)
+    rank = int(os.getenv("DEVICE_ID", 0))
+    group = int(os.getenv("RANK_SIZE", "1"))
+    print("[info] rank: %d group: %d" % (rank, group))
+    ms.context.set_context(device_id=args.device_id)
+
+    np.random.seed(0)
+    ms.set_seed(0)
+
+    model, _ = mindaudio.create_model("FastSpeech2", hps, args.restore, is_train=False)
```
Comment (vigo999) [inline examples/fastspeech2/generate.py:124]: please create fastspeech2 model step by step not using create_model api since this api has been removed for some reason.

File: examples/fastspeech2/preprocess.py [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+)
+parser.add_argument("--device_id", "-i", type=int, default=0)
+parser.add_argument(
+    "--config",
+    "-c",
+    type=str,
+    default="recipes/LJSpeech/tts/fastspeech2/fastspeech2.yaml",
+)
+args = parser.parse_args()
+hps = mindaudio.load_config(args.config)
```
Comment (vigo999) [inline examples/fastspeech2/preprocess.py:33]: same as above load config function. please upload your config.py to examples/fastspeech2

File: examples/fastspeech2/train.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        dt = t - self.t
+        self.t = t
+        print("[loss] %.2f [time] %.2fs" % (loss, dt))
+
+        return loss, cond, scaling_sens
+
+
+def main():
+    args = parse_args()
+    hps = mindaudio.load_config(args.config)
```
Comment (vigo999) [inline examples/fastspeech2/train.py:87]: same problem as above load config

File: examples/fastspeech2/train.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+    model, ckpt = mindaudio.create_model(
```
Comment (vigo999) [inline examples/fastspeech2/train.py:89]: same problem as above for create_model apio

File: examples/wavegrad/reverse.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    old = args.wav.split("/")[-1].replace(".wav", "")
+
+gtwav = wav
+gtmel = feature[0]
+
+feature = ms.Tensor(feature)
+print("old:", old)
+print("feature:", feature.shape)
+
+model, ckpt = mindaudio.create_model("WaveGrad", hps, args.restore, is_train=False)
```
Comment (vigo999) [inline examples/wavegrad/reverse.py:86]: same problem as fastspeech2. please create model step by step with create_model api

File: examples/wavegrad/README.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
+# WaveGrad
+
+MindSpore implementation of [WaveGrad](https://arxiv.org/abs/2009.00713), a diffusion based vocoder model for text-to-speech systems.
```
Comment (vigo999) [inline examples/wavegrad/README.md:3]: remove 'MindSpore implementation of '

File: examples/wavegrad/README.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+1. `pip install -r requirements.txt`
+2. Install [MindSpore](https://www.mindspore.cn/install).
+3. (Optional) Install `mpirun` for distributed training.
+
+## Generate from your data
+
+From wav files:
+
+`python recipes/LJSpeech/tts/wavegrad/reverse.py --restore model_1000000.ckpt --wav LJ010-0142.wav --save results --device_target Ascend --device_id 0 --plot`
```
Comment (vigo999) [inline examples/wavegrad/README.md:25]: please correct scripts path

File: examples/wavegrad/README.md [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+From melspectrograms:
+
+`python recipes/LJSpeech/tts/wavegrad/reverse.py --restore model_1000000.ckpt --mel fs.npy --save results --device_target Ascend --device_id 0 --plot`
```
Comment (vigo999) [inline examples/wavegrad/README.md:29]: scripts path problem

File: examples/wavegrad/README.md [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+`python recipes/LJSpeech/tts/wavegrad/preprocess.py --device_target CPU --device_id 0`
+
+### Step 1 (Train)
+
+#### 1.1 Train on local server
+
+Set up device information:
+```
+export MY_DEVICE=Ascend # options: [Ascend, GPU]
```
Comment (vigo999) [inline examples/wavegrad/README.md:59]: remove MY_DEVICE setting. 
set target device via train.py scripts.

File: examples/wavegrad/README.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+export MY_DEVICE_NUM=8
+```
+
+Other training and model parameters can be set in `base.yaml`.
+
+Train on multiple cards: (each card will have a batch size of hparams.batch_size // MY_DEVICE_NUM)
+```
+nohup mpirun --allow-run-as-root -n $MY_DEVICE_NUM python recipes/LJSpeech/tts/wavegrad/train.py --device_target $MY_DEVICE --is_distributed True --context_mode graph > train_distributed.log &
```
Comment (vigo999) [inline examples/wavegrad/README.md:67]: please correct it as belows:
mpirun --allow-run-as-root -n 8 python train.py --device_target Ascend --is_distributed True --context_mode graph

File: examples/wavegrad/README.md [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+```
+
+Train on 1 card:
+```
+export DEVICE_ID=0
+nohup python train.py --device_target $MY_DEVICE --context_mode graph > train_single.log &
```
Comment (vigo999) [inline examples/wavegrad/README.md:73]: please correct it as below:
python train.py --device_target Ascend --context_mode graph

File: examples/wavegrad/README.md [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+### Implementation details
+
+The interpolation operator in both downsample and upsample blocks are replaced by a simple repeat operator, then divided by repeat factor.
+
+Some additions in UBlock are divided by a constant $\sqrt{2}$ to avoid potential numerical overflow.
+
+### Acknowlegements
+
+Some materials helpful for understanding diffusion models:
+- [youtube](https://www.youtube.com/watch?v=nv-WTeKRLl0)
```
Comment (vigo999) [inline examples/wavegrad/README.md:85]: please fill in [] with video title 
[Learning to Generate Data by Estimating Gradients of the Data Distribution (Yang Song, Stanford)]
ps: this video is for diffusion?? just double confirm

File: examples/wavegrad/README.md [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+- [youtube](https://www.youtube.com/watch?v=HoKDTa5jHvg)
+- [youtube](https://www.youtube.com/watch?v=XCUlnHP1TNM)
+- [math](lilianweng.github.io/posts/2021-07-11-diffusion-models)
+
+Some repositories that inspired this implementation:
+- [lmnt](https://github.com/lmnt-com/wavegrad)
+- [MindAudio](https://github.com/mindspore-lab/mindaudio)
```
Comment (vigo999) [inline examples/wavegrad/README.md:92]: remove mindaudio ref

File: examples/wavegrad/README.md [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+- [FastSpeech2](https://github.com/ming024/FastSpeech2)
+
+Computational Resources:
```
Comment (vigo999) [inline examples/wavegrad/README.md:95]: remove computational resources.

File: examples/wavegrad/preprocess.py [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+        default="recipes/LJSpeech/tts/wavegrad/wavegrad_base.yaml",
+    )
+    parser.add_argument("--device_id", "-i", type=int, default=0)
+    args = parser.parse_args()
+    ms.context.set_context(
+        mode=ms.context.PYNATIVE_MODE,
+        device_target=args.device_target,
+        device_id=args.device_id,
+    )
+    hps = mindaudio.load_config(args.config)
```
Comment (vigo999) [inline examples/wavegrad/preprocess.py:114]: same problem as above 'load config'

File: examples/wavegrad/reverse.py [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+        "--device_target", type=str, default="CPU", choices=("GPU", "CPU", "Ascend")
+    )
+    parser.add_argument("--device_id", "-i", type=int, default=0)
+    parser.add_argument("--save", "-s", type=str, default="results")
+    parser.add_argument("--plot", "-p", type=ast.literal_eval, default=True)
+    parser.add_argument(
+        "--config",
+        "-c",
+        type=str,
+        default="recipes/LJSpeech/tts/wavegrad/wavegrad_base.yaml",
```
Comment (vigo999) [inline examples/wavegrad/reverse.py:27]: correct default config path to examples/xxx

File: examples/wavegrad/reverse.py [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+    parser.add_argument("--wav", "-w", type=str, default="data/LJspeech-1.1/wavs")
+    parser.add_argument("--mel", "-m", type=str, default=None)
+    parser.add_argument("--data_url", default="")
+    parser.add_argument("--train_url", default="")
+    args = parser.parse_args()
+    return args
+
+
+args = parse_args()
+hps = mindaudio.load_config(args.config)
```
Comment (vigo999) [inline examples/wavegrad/reverse.py:40]: 'load config ' problem

File: examples/wavegrad/reverse.py (hunk 14 above)
Comment (vigo999) [inline examples/wavegrad/reverse.py:86]: same problem as 'create_model' api

File: examples/wavegrad/train.py [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+        "--device_target", type=str, default="GPU", choices=("GPU", "CPU", "Ascend")
+    )
+    parser.add_argument(
+        "--context_mode", type=str, default="graph", choices=["py", "graph"]
+    )
+    parser.add_argument(
+        "--config",
+        "-c",
+        type=str,
+        default="recipes/LJSpeech/tts/wavegrad/wavegrad_base.yaml",
```
Comment (vigo999) [inline examples/wavegrad/train.py:29]: correct default config path

File: examples/wavegrad/train.py [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+        overflow = self.process_loss_scale(cond)
+        if not overflow:
+            self.optimizer(grads)
+
+        return loss, cond, scaling_sens
+
+
+def main():
+    args = parse_args()
+    hps = mindaudio.load_config(args.config)
```
Comment (vigo999) [inline examples/wavegrad/train.py:87]: 'load config ' problem

File: examples/wavegrad/train.py [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+    )
+    lr = nn.exponential_decay_lr(
+        hps.learning_rate,
+        0.96,
+        int(ds.get_dataset_size() * hps.num_epochs),
+        int(ds.get_dataset_size()),
+        1000,
+        is_stair=True,
+    )
+    wavegrad, ckpt = mindaudio.create_model(
```
Comment (vigo999) [inline examples/wavegrad/train.py:123]: 'create_model' problem

**Review by vigo999 — CHANGES_REQUESTED** [review]
please check comments

**Review by vigo999 — APPROVED** [review]

**Comment by vigo999** [comment]
pre-commit @DongyaoZhu

**Review by LiTingyu1997 — APPROVED** [review]
